pimd/pim6_mroute_msg.c

```diff
@@ -0,0 +1,191 @@
+				enable ? "MRT6_INIT" : "MRT6_DONE", data, errno,
+				safe_strerror(errno));
+			return -1;
+		}
+	}
+
+	if (enable) {
+		/* Linux and Solaris IPV6_PKTINFO */
+		data = 1;
+		if (setsockopt(pim->mroute_socket, PIM_IPPROTO, IPV6_PKTINFO,
+			       &data, data_len)) {
+			zlog_warn(
+				"Could not set IPV6_PKTINFO on socket fd=%d: errno=%d: %s",
+				pim->mroute_socket, errno,
+				safe_strerror(errno));
+		}
+	}
+
+	setsockopt_so_recvbuf(pim->mroute_socket, 1024 * 1024 * 8);
+
+	flags = fcntl(pim->mroute_socket, F_GETFL, 0);
+	if (flags < 0) {
+		zlog_warn("Could not get flags on socket fd:%d %d %s",
+			  pim->mroute_socket, errno, safe_strerror(errno));
+		close(pim->mroute_socket);
+		return -1;
+	}
+	if (fcntl(pim->mroute_socket, F_SETFL, flags | O_NONBLOCK)) {
+		zlog_warn("Could not set O_NONBLOCK on socket fd:%d %d %s",
+			  pim->mroute_socket, errno, safe_strerror(errno));
+		close(pim->mroute_socket);
+		return -1;
+	}
+
+	if (enable) {
+#if defined linux
+		int upcalls = MRT6MSG_WRMIFWHOLE;
+		opt = MRT6_PIM;
+
+		err = setsockopt(pim->mroute_socket, PIM_IPPROTO, opt, &upcalls,
+				 sizeof(upcalls));
+		if (err) {
+			zlog_warn(
+				"Failure to register for WHOLE and WRONGMIF upcalls %d %s",
+				errno, safe_strerror(errno));
+			return -1;
+		}
+#else
+		zlog_warn(
+			"PIM-SM will not work properly on this platform, until the ability to receive the WHOLEPKT upcall");
+#endif
+	}
+
+	return 0;
+}
+static const char *const mrt6msgtype2str[MRT6MSG_WRMIFWHOLE + 1] = {
+	"<unknown_upcall?>", "NOCACHE", "WRONGMIF", "WHOLEPKT", "WRMIFWHOLE"};
+
+int pim_mroute_msg(struct pim_instance *pim, const char *buf,
+			  size_t buf_size, ifindex_t ifindex)
+{
+	struct interface *ifp;
+	const struct ip6_hdr *ip6_hdr;
+	const struct mrt6msg *msg;
+
+	if (buf_size < (int)sizeof(struct ip6_hdr))
+		return 0;
+
+	ip6_hdr = (const struct ip6_hdr *)buf;
+
+	if ((ip6_hdr->ip6_vfc & 0xf) == 0) {
+		msg = (const struct mrt6msg *)buf;
+
+		ifp = pim_if_find_by_vif_index(pim, msg->im6_mif);
+
+		if (!ifp)
+			return 0;
+		if (PIM_DEBUG_MROUTE) {
+			zlog_debug(
+				"%s: pim kernel upcall %s type=%d ip_p=%d from fd=%d for (S,G)=(%pI4,%pI4) on %s mifi=%d  size=%ld",
+				__func__, mrt6msgtype2str[msg->im6_msgtype],
+				msg->im6_msgtype, ip6_hdr->ip6_nxt,
+				pim->mroute_socket, &msg->im6_src,
+				&msg->im6_dst, ifp->name, msg->im6_mif,
+				(long int)buf_size);
+		}
+
+		switch (msg->im6_msgtype) {
+		case MRT6MSG_WRONGMIF:
+			return pim_mroute_msg_wrongvif(pim->mroute_socket, ifp,
+							msg);
+		case MRT6MSG_NOCACHE:
+			return pim_mroute_msg_nocache(pim->mroute_socket, ifp,
+						       msg);
+		case MRT6MSG_WHOLEPKT:
+			return pim_mroute_msg_wholepkt(pim->mroute_socket, ifp,
+							(const char *)msg);
+		case MRT6MSG_WRMIFWHOLE:
+			return pim_mroute_msg_wrvifwhole(
+				pim->mroute_socket, ifp, (const char *)msg);
+		default:
+			break;
+		}
+	} else {
+		zlog_debug ("Ignoring the ICMPv6 packet in fd %d", 
```

Comment:
this message produces a LOT of noise — please remove

Reply:
(maybe we should add a BPF on the socket so we only get kernel upcalls)

Reply:
Addressed